akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -988,6 +988,37 @@ trait FlowOps[+Out, +Mat] {
    */
   def mapConcat[T](f: Out => IterableOnce[T]): Repr[T] = statefulMapConcat(() => f)
 
+  /**
+   * Transform each input element into a pair of an element of `T` and a state of `S`. The element will be emitted downstream
+   * and the state will be passed to the next step in the chain. The returned `T` MUST NOT be `null` as it is illegal
+   * as stream element - according to the Reactive Streams specification. For stateless variant see [[FlowOps.map]].
+   *
+   * The initial state is created using the given function `create` when the flow is materializing.
+   *
+   * The `onComplete` function is called only once when the upstream or downstream finished, You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam S the type of the state
+   * @tparam T the type of the output elements
+   * @param create a function that creates the initial state
+   * @param f a function that transforms the upstream element and the state into a pair of next state and output element
+   * @param onComplete a function that transforms the ongoing state into an optional output element
+   */
+  def statefulMap[S, T](create: () => S)(
+      f: (S, Out) => (S, T),
+      onComplete: S => T = StatefulMap.IgnoreStateFunction.asInstanceOf[S => T]): Repr[T] =
```

Comment:
I think it should be a bad idea to make use of default parameter here.

Reply:
@otto-dev @johanandren I think it still can keep the ability for onComplete to decide whether or not a final element should be emitted after making the `onComplete` return a `T`, but that should involves a special function value seems not good.

Reply:
_
![image](https://user-images.githubusercontent.com/501740/162801048-c402d938-7b5c-41c1-b8fe-4997d404b9f7.png)
_

Reply:
if we make the `onComplete` return a `T` and not using a special function to indicate the optional emmit for the final value, then there will always an additional demand is need to finish the stage.

Reply:
Yes, no default parameters, they are basically impossible to evolve in binary compatible ways.

What I meant with making it non optional is that a user can turn it optional through composition:

```
source.statefulMap[State, Option[T]](...,...,...).collect { case Some(t) => t }
```

On the other hand that will mean an extra `Some` allocated for each normal element, so maybe it was a bad idea, let's keep the optional `onComplete` for efficiency.